Make measurement resubmission faster

bassosimone opened the pull request:
# Description

This diff takes advantage of recents changes in mk* libraries [1] to implement faster measurements resubmission [2] for Android.

We vendor mkcurl v0.12.0, mkbouncer v0.2.0, and a development version of mkcollector [3], to have code able to reuse existing open HTTP connections.

We implement a new MKReporter class that uses new code in mkcollector such that we open and close reports only when the measurements we are submitting are clearly part of different reports. That is, the app does not need to concern itself with this problem.

We do not anticipate specific issues in integrating with the OONI app for Android because resubmission is sync code and therefore there are no significant concerns around keeping objects alive.

# Scope of the review

Please, ignore the files in `.../cpp/vendor`. They are copied from other repositories where there is very high code coverage and they have also already been reviewed. Focus on the newly added Java and JNI C++ code and specifically on:

- [ ] correctness
- [ ] whether the API seems appropriate
- [ ] whether more tests are needed

# Roll out plan

- [x] I will tag a release of android-libs out of this branch. *This means that the branch will qualify as being merged without squashing*. The reason why this operation is safe is because the only change is this new `MKReporter` interface and all the rest is unchanged. The release message will clarify that this interface is still unstable and subject to changes.

- [x] Then, I will create a branch in the https://github.com/ooni/probe-android repository using this new version and I will make sure that we can smoothly transition to the new API.

- [x] When https://github.com/measurement-kit/mkcollector/pull/25 is approved, I will tag a new stable release of mkcollector and vendor that into this repository.

- [x] Once we have updated the probe-ios codebase, I will remove the previous interface for submitting measurements because at this point that has become obsolete.

- [x] This will open up the opportunity of upgrading this repository to MK v0.10.5, where a number of vendored files that where used by resubmission have now been removed.

# Notes

1. measurement-kit/measurement-kit#1867

2. ooni/probe#861

3. measurement-kit/mkcollector#25

bassosimone committed:
Make measurement resubmission faster

Part of https://github.com/ooni/probe/issues/861

bassosimone commented on android-libs/src/androidTest/java/io/ooni/mk/androidTest/MKReporterTest.java:
```diff
@@ -0,0 +1,39 @@
+        System.out.println("updatedReportID     : " + results.getUpdatedReportID());
+        System.out.println("serializedStats     : " + results.getSerializedStats());
+        Truth.assertThat(results.isGood()).isTrue();
+    }
```
We probably need to add more tests before we can bless this PR as ready for review.

bassosimone commented on android-libs/src/main/cpp/reporter.cpp:
```diff
@@ -0,0 +1,106 @@
+  }
+  auto results = new mk_reporter_results{};
+  if (!mkall_string_java2cxx(env, measurement, results->measurement)) {
+    return 0L;  // Exception already pending
```
Here we're leaking `results`. We should use a unique pointer.

bassosimone commented on android-libs/src/main/cpp/reporter.cpp:
```diff
@@ -0,0 +1,106 @@
+#define XX(name_) doc[#name_] = stats.name_;
+    MKCOLLECTOR_REPORTER_STATS_ENUM(XX)
+#undef XX
+    results->stats = doc.dump();
```
Mention that this is very unlikely to throw.

bassosimone commented on android-libs/src/main/cpp/reporter.cpp:
```diff
@@ -0,0 +1,106 @@
+#undef XX
+    results->stats = doc.dump();
+  }
+  return reinterpret_cast<jlong>(results);
```
Justify why using `reinterpret_cast<>` here.

bassosimone replied:
According to [cppreference,com](https://en.cppreference.com/w/cpp/language/reinterpret_cast):

> [...] the reinterpret_cast expression does not compile to any CPU instructions (except when converting between integers and pointers or on obscure architectures where pointer representation depends on its type).
>
> Only the following conversions can be done with reinterpret_cast:
> [...]
> 2) A pointer can be converted to any integral type large enough to hold all values of its type [...]
>
> 3) A value of any integral or enumeration type can be converted to a pointer type. A pointer converted to an integer of sufficient size and back to the same pointer type is guaranteed to have its original value [...]

So, here we're casting with reinterpret cast the pointer to `jlong` (case 2). Then, when we are to use the pointer, we cast back (case 3). We're good.

bassosimone reviewed (commented):
Made a list of improvements before we can declare this ready for review

bassosimone committed:
script/common/javah: cope with missing javah in JDK 12

I've just upgraded to JDK 12 where there's no `javah`. There is
`javac -h <dir>` though and we can use it.

The upgrade to JDK 12 also caused a number of annotations to
be added, to silence the linter. Specifically, with Android we
still use and require the finalizer.

bassosimone committed:
mkall_util.h: add diagnostic code for string misencoding

Reference bug: https://github.com/measurement-kit/measurement-kit/issues/1876

bassosimone committed:
Release v0.10.4-2-android.5

This release is like v0.10.4-2-android.4 except that it also adds diagnostic
code to investigate https://github.com/measurement-kit/measurement-kit/issues/1876.

bassosimone committed:
mkall_util.h: add diagnostic code for string misencoding (try 2)

A better version of c95d495a3606c9e0a8a9bbceb8d3a54524dbb862 that
hopefully this time should log something.

> By default, the Android system sends stdout and stderr (System.out and
> System.err) output to /dev/null. In processes that run the Dalvik VM, you
> can have the system write a copy of the output to the log file.

Source: https://stackoverflow.com/a/36561699/4354461

bassosimone committed:
Release v0.10.4-2-android.6

Like 0.10.4-2-android.5 but this time the diagnostic code to
reveal what is the culprit string should work (TM).

bassosimone committed:
mkall_util.h: add diagnostic code for string misencoding (try 3)

Make diagnostic code unconditional. We'll need to remove it for sure
but this should really be invoked this time.

I didn't read the stack trace in depth and assumed that the code
was going to return an exception. Instead it was aborting right into
env->NewStringUTF therefore there was no point in adding actions
after that function call (as we did in try 1 and try 2).

bassosimone committed:
Release v0.10.4-2-android.7

Like v0.10.4-2-android.6 but with a hopefully better strategy to gather
useful data to understand e measurement-kit/measurement-kit#1876.

bassosimone committed:
mkall_util.h: add diagnostic code for string misencoding (try 4)

We're clearly starting to have a problem here. While recent changes
enabled faster development, passing through this is clearly the next
bottleneck to be removed. I'll think about this later.

For now, print also the length of the string. It seems the android
logger is truncating it, and that will allow to confirm.

bassosimone committed:
Release v0.10.4-2-android.8

Continuing to hunt the measurement-kit/measurement-kit#1876 and exposing
how the current code structure is too rigid because that has taken
several "releases" and there is no clear fix in sight yet.

bassosimone committed:
mkall_util.h: add diagnostic code for string misencoding (try 6)

Now we know that the string is indeed truncated by the Android logcat and
we want to know, as the next step, whether it's valid UTF-8.

bassosimone committed:
Release v0.10.4-2-android.9

lorenzoPrimi reviewed (commented):
Reviewed the .java files and they all seems ok.
More tests, especially failing tests are needed

bassosimone committed:
android-libs: adapt to using MK v0.10.5

Mainly, get rid of the collector API, adapt the geoiplookup API to
use C++ directly, and remove unneded mkcurl vendoring.

bassosimone committed:
Release v0.10.5-android.1

bassosimone committed:
Makefile: make sure we're building the right version

This is a double check. We need to be sure the version we're building
is the same that is installed using homebrew.

bassosimone committed:
reporter.cpp: use a unique pointer

This plugs leaks on the error path because the unique pointer
destroys the allocated memory on destruction.

Extract the value from the unique pointer on success so that we
can transfer its ownership to Java code.

See https://github.com/measurement-kit/android-libs/pull/78#discussion_r307742897

bassosimone committed:
reporter.cpp: document why a statement cannot throw

See https://github.com/measurement-kit/android-libs/pull/78#discussion_r307743286

bassosimone committed:
mkall_util.h: remove debugging leftovers

Such code has been useful to diagnose the issue and to formulate a
strategy to mitigate it. So we can now remove it.

bassosimone committed:
MKReportTest.java: start testing multiple submissions

bassosimone committed:
Makefile: bump the version code

bassosimone commented:
Additional issues spotted while testing this PR:

- [ ] the NDT test does not run because it cannot load verify locations

- [ ] the orchestra test does not work

- [ ] we need to make sure we use the latest version of vendored deps and add a note about this inside of the README.md file

- [ ] we need to change unit tests to enforce in more cases that they MUST work

- [ ] also web connectivity fails like NDT because of wrong verify locations

bassosimone committed:
s/c/javah: clean up *.class files when done

bassosimone committed:
Upgrade to measurement-kit/mkcollector@v0.6.1

While there add a note to README.md that we need to update the
vendored dependencies when we're uprading.

bassosimone committed:
androidTest: make sure all tests have correct resources

bassosimone commented on android-libs/src/androidTest/java/io/ooni/mk/androidTest/MKReporterTest.java:
```diff
@@ -0,0 +1,58 @@
+        System.loadLibrary("measurement_kit");
+    }
+
+    static private String baseMeasurement = "{\"data_format_version\":\"0.2.0\",\"id\":\"9858087b-15b9-4913-a87e-2520f85393b2\",\"input\":null,\"input_hashes\":[],\"measurement_start_time\":\"2019-07-26 10:39:09\",\"options\":[],\"probe_asn\":\"AS0\",\"probe_cc\":\"ZZ\",\"probe_city\":null,\"probe_ip\":\"127.0.0.1\",\"report_id\":\"20190726T103909Z_AS0_KTgS3f6IgfdSuCoqQE1A31u4AtjpbO6Ywj0LPkn7c4mT ...
```
This is extremely ugly. We should instead use Gson also here.

bassosimone commented on android-libs/src/androidTest/java/io/ooni/mk/androidTest/MKWebConnectivityTest.java:
```diff
@@ -3,26 +3,33 @@
+        settings.options.geoip_country_path = MKResourcesManager.getCountryDBPath(context);
         settings.inputs = new ArrayList<String>();
         settings.inputs.add("https://www.torproject.org/");
         settings.inputs.add("https://x.org/");
```
We probably want to test the URL that was causing https://github.com/measurement-kit/measurement-kit/issues/1877

bassosimone commented on android-libs/src/main/cpp/geoiplookup.cpp:
```diff
@@ -2,65 +2,88 @@
-               mk_geoiplookup_response_logs_at,
-               mk_geoiplookup_response_t)
+               [](auto r) noexcept { return r->logs.size(); },
+               [](auto r, size_t i) noexcept { return r->logs[i].c_str(); },
```
This and other places could possibly return UTF-8 strings not compatible with MUTF-8. The proper thing to do would be to encoded to base64 if the string is not MUTF-8 compatible.

bassosimone replied:
Also, perhaps we can take advantage of the availability of a C++ string and use that directly.

bassosimone replied:
We should create an issue to do this later. This diff is too large.

bassosimone replied:
https://github.com/measurement-kit/android-libs/issues/80

bassosimone commented on android-libs/src/main/cpp/mkall_util.h:
```diff
@@ -4,7 +4,10 @@
 #ifndef SRC_MKALL_HEADERS_MKALL_UTIL_H
 #define SRC_MKALL_HEADERS_MKALL_UTIL_H
 
+#include <android/log.h>
```
This include seems unnecessary.

bassosimone commented on android-libs/src/main/cpp/mkall_util.h:
```diff
@@ -184,6 +187,23 @@
     cxx_func((cxx_type *)handle, value);                \
   }
 
+// mkall_string_java2cxx copies the @p in string to the @p out string.
```
We should document the shortcomings of this function vis a vis MUTF-8.

bassosimone replied:
#80

bassosimone commented on android-libs/src/main/cpp/reporter.cpp:
```diff
@@ -0,0 +1,111 @@
+    return 0L;
+  }
+  std::string software_name;
+  std::string software_version;
```
Using variables that differ only because they're snake case rather than camel case is really bad.

bassosimone commented on android-libs/src/main/java/io/ooni/mk/MKReporterTask.java:
```diff
@@ -0,0 +1,47 @@
+        SetCABundlePath(handle, caBundlePath);
+    }
+
+    /** submit submites a measurement and returns the results. This method will
```
`submites` => `submits`

bassosimone commented on android-libs/src/main/java/io/ooni/mk/MKReporterTask.java:
```diff
@@ -0,0 +1,47 @@
+
+    /** submit submites a measurement and returns the results. This method will
+     * automatically discover a collector, if none is specified. */
+    public MKReporterResults submit(String measurement, long timeout) {
```
We should give this function a less surprising name (the private native name it's fine).

bassosimone reviewed (commented):
Pointed out a bunch of places where we can improve

bassosimone committed:
androidTest: rationalize imports

bassosimone committed:
androidTest: refactor and add some checks

While there, make sure that MKOrchestraTest.java actually runs.

Also, remove the UnitTestSuite.java because we don't use it.

bassosimone committed:
mkall_util.h: remove unneeded import

See https://github.com/measurement-kit/android-libs/pull/78#discussion_r318219026

bassosimone committed:
reporter.cpp: use better variable names

See https://github.com/measurement-kit/android-libs/pull/78#discussion_r318219747

bassosimone committed:
MKReporterTask.java: fix typo

See https://github.com/measurement-kit/android-libs/pull/78#discussion_r318221115

bassosimone committed:
MKReporterTask: better name for submit

See https://github.com/measurement-kit/android-libs/pull/78#discussion_r318221388

bassosimone committed:
androidTest: give non ambiguous name to settings class

bassosimone committed:
androidTest: better testts for NDT and WebConnectivity

bassosimone committed:
androidTest: add unit test for fanfiction.net

The test is passing locally. It means we've really fixed the
original but at https://github.com/ooni/probe-android/issues/248.

bassosimone committed:
MKReporterTest: use Gson

See https://github.com/measurement-kit/android-libs/pull/78#discussion_r318216967

bassosimone committed:
MKReporterTest: verifies events match expectation

bassosimone commented:
@lorenzoPrimi can you take another look? The diff since you've last viewed is https://github.com/measurement-kit/android-libs/compare/7565847...387a1d3
